packages/core/admin/ee/server/services/review-workflows/assigned-content-types.js

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const { mapAsync } = require('@strapi/utils');
+const { difference } = require('lodash/fp');
+const { getService } = require('../../utils');
+const { WORKFLOW_MODEL_UID } = require('../../constants/workflows');
+
+module.exports = () => ({
+  async create({ assigned, defaultStage }) {
+    await mapAsync(assigned, async (uid) => {
+      const swappingFromWorkflow = await this.getAssignedWorkflow(uid);
+      // Content type is already assigned to a workflow, swap it
+      if (swappingFromWorkflow) {
+        return this.swapAssignedContentType(swappingFromWorkflow, defaultStage, uid);
+      }
+      // Content type is not assigned to a workflow, init it
+      return this.initEntityStages(uid, defaultStage);
+    });
+  },
+
+  async update({ srcAssigned, destAssigned, defaultStage }) {
+    // Find difference between old assignedContentTypes and new ones, using lodash fp
+    const { created, deleted } = diffAssignedContentTypes(srcAssigned, destAssigned);
+
+    await mapAsync(created, async (uid) => {
+      const swappingFromWorkflow = await this.getAssignedWorkflow(uid);
+      // Content type is already assigned to a workflow, swap it
+      if (swappingFromWorkflow) {
+        return this.swapAssignedContentType(swappingFromWorkflow, defaultStage, uid);
+      }
+      // Content type is not assigned to a workflow, init it
+      return this.initEntityStages(uid, defaultStage);
+    });
+
+    await mapAsync(deleted, async (uid) => {
+      await getService('stages').deleteAllEntitiesStage(uid, {});
+    });
+  },
+
+  /**
+   * Swapping involves:
+   *  - Updating entity stages of the content type
+   *  - Updating the contentTypes of the previous workflow
+   * @param {Workflow} srcWorkflow
+   * @param {Workflow.Stage} stage
+   * @param {string} uid
+   */
+  async swapAssignedContentType(srcWorkflow, stage, uid) {
+    // Update entity stages of the content type
+    await getService('stages').updateAllEntitiesStage(uid, { toStageId: stage.id });
+    // Update assignedContentTypes of the previous workflow
+    await strapi.entityService.update(WORKFLOW_MODEL_UID, srcWorkflow.id, {
+      data: {
+        contentTypes: srcWorkflow.contentTypes.filter((ct) => ct !== uid),
+      },
+    });
+  },
+
+  async initEntityStages(uid, stage) {
+    return getService('stages').updateEntitiesStage(uid, {
+      fromStageId: null,
+      toStageId: stage.id,
+    });
+  },
+
+  async getAssignedWorkflow(uid, opts) {
+    // TODO: Improve filter
+    const workflows = await getService('workflows').find({
+      ...opts,
+      filters: { contentTypes: { $contains: [uid] } },
+    });
+    return workflows.length > 0 ? workflows[0] : null;
```

Comment:
Likewise should this only ever be of length 1? Should we enforce that here and throw an error otherwise?

Reply:
I would not throw in a "get" action, as it might make the admin unresponsive to changes and making it impossible to fix.

Imo the check needs to be in the create and update, but probably we could add some sanity checks somewhere else 🤔  maybe on startup?